v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts

```diff
@@ -119,33 +122,46 @@ export async function* testReporter(
     for (const [index, failure] of failures.entries()) {
       yield `\n${chalk.bold(chalk.red(`Failure (${index + 1})`))}: ${failure.name}\n`;
 
-      if (
-        failure.reason !== undefined &&
-        failure.reason !== null &&
-        failure.reason !== ""
-      ) {
-        yield `Reason: ${chalk.grey(failure.reason)}\n`;
+      const stackTrace = failure.stackTrace();
+
+      let reason: string | undefined;
+      if (stackTrace?.kind === "StackTrace") {
+        reason = getMessageFromLastStackTraceEntry(stackTrace.entries[stackTrace.entries.length - 1]);
+      }
+      if (reason === undefined || reason === "") {
+        reason = failure.reason ?? "Unknown error";
       }
 
-      const stackTrace = failure.stackTrace();
-      // TODO handle `UnexpectedError` and `UnsafeToReplay` variants
-      if (
-        stackTrace !== undefined &&
-        stackTrace !== null &&
-        stackTrace.kind === "StackTrace" &&
-        stackTrace.entries.length > 0
-      ) {
-        const stackTraceStack: string[] = [];
-        for (const entry of stackTrace.entries.reverse()) {
-          const callsite = encodeStackTraceEntry(entry);
-          if (callsite !== undefined) {
-            stackTraceStack.push(`  at ${callsite.toString()}`);
+      yield `Reason: ${chalk.grey(reason)}\n`;
+
+      switch (stackTrace?.kind) {
+        case "StackTrace":
+          const stackTraceStack: string[] = [];
+          for (const entry of stackTrace.entries.reverse()) {
+            const callsite = encodeStackTraceEntry(entry);
+            if (callsite !== undefined) {
+              stackTraceStack.push(`  at ${callsite.toString()}`);
+            }
           }
-        }
 
-        if (stackTraceStack.length > 0) {
-          yield `${chalk.grey(stackTraceStack.join("\n"))}\n`;
-        }
+          if (stackTraceStack.length > 0) {
+            yield `${chalk.grey(stackTraceStack.join("\n"))}\n`;
+          }
+          break;
+        case "UnexpectedError":
+          yield `Stack Trace Warning: ${chalk.grey(stackTrace.errorMessage)}\n`;
```

Comment:
This used to throw before. I think it's ok to turn it into a warning, but wanted to make sure we're all on the same page about this.

Reply:
Is this an unexpected error in the stack traces generation? Or when? Maybe the name can be more explicit.

Reply:
Yes, that's during stack trace generation.